### PR TITLE
adding documentation from issues in Appendix C

### PR DESCRIPTION
**Pull request by larahabashy**

### CODE REVIEW

File: documentation.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+- CLAHE
+- Downscale
+- Equalize 
+- Hue Saturation (may not apply if no color channels)
+- ISO noise 
+- Invert Image
+- Random Birghtness
+- Random Contrat 
+- Random Snow
+- Shift Scale rotate
```
Comment (jraza19) [inline documentation.md]: I think most of these details can be taken out in terms of listing them here. We can that we initially used a [web interface](https://albumentations-demo.herokuapp.com/) provided by the albumentations team that allowed us to see what the different transformations looked like and based on that, we chose some transformations to test in our pipeline. The final model was chosen based on accuracy score.

File: documentation.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+3. Next, run the following command:
+`autoalbument-search --config-dir ~/mds/capstone/capstone-gdrl-lipo/experiments`
+
+4. Finally, you get a [json file](https://github.com/UBC-MDS/capstone-gdrl-lipo/tree/auto_exp1/experiments/outputs/2021-05-11/16-45-11/policy)  that can them be loaded in and applied to your transformation. See the transformation exploration [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/auto_exp1/notebooks/autoalbument.ipynb).
+
+### Loss Functions
+
+A loss function, also known as cost function or error function, is a function that maps a set of parameter values for the network onto a scalar value. The scalar value is an indication of how well the parameters are comleting their assigned tasks. In optimization problems, we seek to minimize the loss function.
```
Comment (jraza19) [inline documentation.md]: Spelling: *completing

File: documentation.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+The following functions were considered for the choice of the loss function used in the convolutional neural network model:
+
+1. BCEWithLogitsLoss
+2. NLLLoss
+3. Sum of Log Loss
+4. Hinge
+5. The mean absolute error 
+
+- BCEWithLogitsLoss: Cross-Entropy loss [nn.BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html#torch.nn.BCEWithLogitsLoss) combines a Sigmoid layer and the BCELoss in one single class. This loss function in often used in classification problems as a measure of reconstruction error. It uses log loss, log loss for logistic regression is a special case for cross-entropy loss for multi-class classification.
```
Comment (jraza19) [inline documentation.md]: Suggestion: BCEWithLogitsLoss: Cross-Entropy loss [nn.BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html#torch.nn.BCEWithLogitsLoss) combines a Sigmoid layer and the BCELoss in one single class. This loss function in often used in classification problems as a measure of reconstruction error. It uses log loss, which for logistic regression is a special case for cross-entropy loss for multi-class classification.

File: documentation.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+The model variants that were considered are documented below, along with some notes thought to be relevant.
+
+- VGG16: It makes the improvement over AlexNet by replacing large kernel-sized filters (**11** and **5** in the first and second convolutional layer, respectively) with multiple 3×3 kernel-sized filters one after another.
+- DenseNet121: DenseNet121 proved to be the best performing model given the Lipohypertrophy data. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/model_decision_making.ipynb). 
+- ResNet50
+- InceptionV3
+
+#### Batch Normalization
+Ensure all models have batch normalization implemented. This will standardize the inputs of a network improving the overall efficiency by reducing the number of epochs required to train for a given model.
```
Comment (jraza19) [inline documentation.md]: Start with "We ensured all models..."

File: documentation.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+#### Dropout Layer
+The implementation of dropout layers within the model's architecture to reduce the generalizable error were considered. This is due to the fact that a dropout layer with randomly with probably equal to the dropout rate, drop nodes from a network between layers. Given the Lipohypertrophy data, the dropout layers experiments with varying dropout rates have shown no success. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/densemodels-ax-dropout-layers.ipynb).
```
Comment (jraza19) [inline documentation.md]: Suggestion: "The implementation of dropout layers within the model's architecture to reduce the generalizable error were considered. This is due to the fact that a dropout layer with random probability equal to the dropout rate, drop nodes from a network between layers. However, the dropout layers experiment with varying dropout rates did not show any success for our dataset. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/densemodels-ax-dropout-layers.ipynb)"

File: documentation.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+#### Dropout Layer
+The implementation of dropout layers within the model's architecture to reduce the generalizable error were considered. This is due to the fact that a dropout layer with randomly with probably equal to the dropout rate, drop nodes from a network between layers. Given the Lipohypertrophy data, the dropout layers experiments with varying dropout rates have shown no success. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/densemodels-ax-dropout-layers.ipynb). 
+
+#### Recall
+To reduce the generalization error, consider varying the pos_weight argument in the loss function. Increasing pos_weight > 1 means there is a heavier penalization (loss) on positives in an attempt to reduce false negatives (a true positive where the model predicts is negative) and improve recall. This experumentation is also used to combat the class imbalance issue. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/pos-weight-exploration.ipynb).
```
Comment (jraza19) [inline documentation.md]: Suggestion: "To reduce the generalization error, we considered varying the pos_weight argument in the loss function. Increasing pos_weight > 1 means there is a heavier penalization (loss) on positives in an attempt to reduce false negatives (a true positive where the model predicts is negative) and improve recall. This experumentation is also used to combat the class imbalance issue. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/pos-weight-exploration.ipynb)".

File: documentation.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+
+### Hyperparameter Optimization - Bayesian Optimization with Ax
+
+The general idea with the Bayesian method is that not every possible parameterization in the defined parameter space is explored.
+The process tunes parameters in a few iterations by building a surrogate model. A surrogate model is a probablistic model that uses an acquisition function to direct the next sample to make up configurations where an improvement over the current best parameterization is likely. Bayesian Optimization with Ax, a recently developed package for optimization by Facebook, relies on information from previous trials to propose better hyperparameters in the next evaluation. The Gaussian process is as follows.
```
Comment (jraza19) [inline documentation.md]: add a : colon at the end

File: documentation.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+Fit new surrogate model and repeat process.
+
+As a way to guide the surrogate model in the prediction of the next parameter configuration to explore, aquisition functions are utilizied. 
+- Three common examples include:
+ 1. Probability of Improvement (PI).
+ 2. Expected Improvement (EI).
+ 3. Lower Confidence Bound (LCB)
+ 
+The most common technique used in Bayesian Optimization is the second choice: Expected Improvement. As such, that was used in the processes experimented [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/densenet-optimized.ipynb).
```
Comment (jraza19) [inline documentation.md]: Suggestion: "The most common technique used in Bayesian Optimization is the second choice: Expected Improvement. As such, that was used for our [experiments](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/densenet-optimized.ipynb)".

**Review by jraza19 — APPROVED** [review]
Great work @larahabashy !

**Review by jraza19 — APPROVED** [review]
